### PR TITLE
Allow raptor shards table be scaned faster

### PR DESCRIPTION
**Pull request by hellium01**
Push down the predicate to mysql to increase the speed of metadata
query as we already know the table_id in advance.

### CODE REVIEW

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java [hunk 1]
```diff
@@ -58,13 +59,15 @@ public static PreparedStatement create(
             List<String> columnNames,
             List<Type> types,
             Set<Integer> uuidColumnIndexes,
-            TupleDomain<Integer> tupleDomain)
+            TupleDomain<Integer> tupleDomain,
+            Optional<String> additionalFixedCondition
+            )
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java]: Formatting

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java [hunk 2]
```diff
@@ -84,9 +87,11 @@ private static String getWhereClause(
             List<String> columnNames,
             List<Type> types,
             Set<Integer> uuidColumnIndexes,
-            List<ValueBuffer> bindValues)
+            List<ValueBuffer> bindValues,
+            Optional<String> additionalCondition
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java]: Formatting

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java [hunk 3]
```diff
@@ -272,7 +273,9 @@ private ResultSet getNextResultSet()
                     columnNames,
                     TYPES,
                     ImmutableSet.of(getColumnIndex(SHARD_METADATA, SHARD_UUID)),
-                    tupleDomain);
+                    tupleDomain,
+                    Optional.of(format("shards.table_id = %d", tableId))
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java]: Formatting

File: presto-raptor/src/test/java/com/facebook/presto/raptor/systemtables/TestShardMetadataRecordCursor.java [hunk 4]
```diff
@@ -146,6 +146,11 @@ public void testSimple()
                 new MaterializedRow(DEFAULT_PRECISION, schema, table, utf8Slice(uuid3.toString()), null, 300L, 30L, 3L, null, null));
 
         assertEquals(actual, expected);
+        try (RecordCursor cursor = new ShardMetadataSystemTable(dbi).cursor(null, SESSION, TupleDomain.all())) {
```
Comment (electrum) [inline presto-raptor/src/test/java/com/facebook/presto/raptor/systemtables/TestShardMetadataRecordCursor.java]: Add newline

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java [hunk 5]
```diff
@@ -98,6 +103,9 @@ private static String getWhereClause(
             Type type = types.get(index);
```
Comment (AlekseiS) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java]: on line 99 tuple domain can be none. I think you need a separate condition to verify it

**Comment by hellium01** [comment]
Updated to address the comments

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java [hunk 6]
```diff
@@ -272,7 +273,9 @@ private ResultSet getNextResultSet()
...
+                    );
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/ShardMetadataRecordCursor.java]: This goes on previous line

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java [hunk 7]
```diff
@@ -84,19 +86,22 @@ private static String getWhereClause(
...
         }
 
-        ImmutableList.Builder<String> conjunctsBuilder = ImmutableList.builder();
-        Map<Integer, Domain> domainMap = tupleDomain.getDomains().get();
-        for (Map.Entry<Integer, Domain> entry : domainMap.entrySet()) {
-            int index = entry.getKey();
-            String columnName = columnNames.get(index);
-            Type type = types.get(index);
-            conjunctsBuilder.add(toPredicate(index, columnName, type, entry.getValue(), uuidColumnIndexes, bindValues));
+        if (additionalFixedCondition.isPresent()) {
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java]: This "if present then get" pattern can be replaced with `.ifPresent`

File: presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java [hunk 8]
```diff
@@ -84,19 +86,22 @@ private static String getWhereClause(
...
             List<Type> types,
             Set<Integer> uuidColumnIndexes,
-            List<ValueBuffer> bindValues)
+            List<ValueBuffer> bindValues,
+            Optional<String> additionalFixedCondition)
     {
-        if (tupleDomain.isNone()) {
-            return "";
+        ImmutableList.Builder<String> conjunctsBuilder = ImmutableList.builder();
+        if (!tupleDomain.isNone()) {
```
Comment (electrum) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/systemtables/PreparedStatementBuilder.java]: This refactoring makes it difficult to see the real change. Please split this out into a separate commit.

**Comment by electrum** [comment]
There's actually a much simpler way: https://github.com/prestodb/presto/pull/8083
